### ISSUE TITLE
rails new --dev not working

### ISSUE THREAD

**Issue by zzak** [rails/rails#47435]
`rails new --dev` is not working as expected for me, it generates the correct gemfile then tries to overwrite it with a "--edge" style gemfile.

### Steps to reproduce

```ruby
git clone https://github.com/rails/rails
cd rails
bundle install
bundle exec railties/exe/rails new --dev ~/code/railsdev
```

### Expected behavior

Generates a Rails app pointing at my rails checkout.

### Actual behavior

<details>
<summary>A conflict occurs with the Gemfile:</summary>

```diff
--- /Users/zzak/code/railsdev/Gemfile	2023-02-20 14:15:32.909345853 +0900
+++ /Users/zzak/code/railsdev/Gemfile20230220-42574-u0vnum	2023-02-20 14:15:35.898757043 +0900
@@ -1,4 +1,71 @@
 source "https://rubygems.org"
 git_source(:github) { |repo| "https://github.com/#{repo}.git" }
-# Use local checkout of Rails
-gem "rails", path: "/Users/zzak/code/rails"
+
+ruby "3.3.0.dev"
+
+# Bundle edge Rails instead: gem "rails", github: "rails/rails", branch: "main"
+gem "rails", "~> 7.1.0.alpha"
+
+# The original asset pipeline for Rails [https://github.com/rails/sprockets-rails]
+gem "sprockets-rails"
+
+# Use sqlite3 as the database for Active Record
+gem "sqlite3", "~> 1.4"
+
+# Use the Puma web server [https://github.com/puma/puma]
+gem "puma", ">= 5.0"
+
+# Use JavaScript with ESM import maps [https://github.com/rails/importmap-rails]
+gem "importmap-rails"
+
+# Hotwire's SPA-like page accelerator [https://turbo.hotwired.dev]
+gem "turbo-rails"
+
+# Hotwire's modest JavaScript framework [https://stimulus.hotwired.dev]
+gem "stimulus-rails"
+
+# Build JSON APIs with ease [https://github.com/rails/jbuilder]
+gem "jbuilder"
+
+# Use Redis adapter to run Action Cable in production
+# gem "redis", ">= 4.0.1"
+
+# Use Kredis to get higher-level data types in Redis [https://github.com/rails/kredis]
+# gem "kredis"
+
+# Use Active Model has_secure_password [https://guides.rubyonrails.org/active_model_basics.html#securepassword]
+# gem "bcrypt", "~> 3.1.7"
+
+# Windows does not include zoneinfo files, so bundle the tzinfo-data gem
+gem "tzinfo-data", platforms: %i[ windows jruby ]
+
+# Reduces boot times through caching; required in config/boot.rb
+gem "bootsnap", require: false
+
+# Use Active Storage variants [https://guides.rubyonrails.org/active_storage_overview.html#transforming-images]
+# gem "image_processing", "~> 1.2"
+
+group :development, :test do
+  # See https://guides.rubyonrails.org/debugging_rails_applications.html#debugging-with-the-debug-gem
+  gem "debug", platforms: %i[ mri windows ]
+end
+
+group :development do
+  # Use console on exceptions pages [https://github.com/rails/web-console]
+  gem "web-console"
+
+  # Add speed badges [https://github.com/MiniProfiler/rack-mini-profiler]
+  # gem "rack-mini-profiler"
+
+  # Speed up commands on slow machines / big apps [https://github.com/rails/spring]
+  # gem "spring"
+
+  gem "error_highlight", ">= 0.4.0", platforms: [:ruby]
+end
+
+group :test do
+  # Use system testing [https://guides.rubyonrails.org/testing.html#system-testing]
+  gem "capybara"
+  gem "selenium-webdriver"
+  gem "webdrivers"
+end
```
</details>

<details>
<summary>If I hit "no", I get an error about missing bootsnap, and probably other gems</summary>

```
Use `bundle info [gemname]` to see where a bundled gem is installed.
         run  bundle binstubs bundler
       rails  importmap:install
/Users/zzak/code/railsdev/config/boot.rb:4:in `require': cannot load such file -- bootsnap/setup (LoadError)
	from /Users/zzak/code/railsdev/config/boot.rb:4:in `<top (required)>'
	from bin/rails:3:in `require_relative'
	from bin/rails:3:in `<main>'
       rails  turbo:install stimulus:install
/Users/zzak/code/railsdev/config/boot.rb:4:in `require': cannot load such file -- bootsnap/setup (LoadError)
	from /Users/zzak/code/railsdev/config/boot.rb:4:in `<top (required)>'
	from bin/rails:3:in `require_relative'
	from bin/rails:3:in `<main>'
```
</details>

<details>
<summary>If I hit "yes", I get an error that bundler cannot find rails version `~> 7.1.0.alpha`</summary>

```
Could not find gem 'rails (~> 7.1.0.alpha)' in rubygems repository https://rubygems.org/ or installed locally.

The source contains the following gems matching 'rails':
```

</details>

<details>
<summary>This also looks similar to the output I'm seeing in the railties test</summary>

```
bin/test test/generators/app_generator_test.rb

/Users/zzak/.rubies/ruby-3.3.0/lib/ruby/3.3.0+0/bundler/resolver.rb:293:in `raise_not_found!': Could not find gem 'rails (~> 7.1.0.alpha)' in locally installed gems. (Bundler::GemNotFound)
	from /Users/zzak/.rubies/ruby-3.3.0/lib/ruby/3.3.0+0/bundler/resolver.rb:346:in `block in prepare_dependencies'
	from /Users/zzak/.rubies/ruby-3.3.0/lib/ruby/3.3.0+0/bundler/resolver.rb:331:in `each'
	from /Users/zzak/.rubies/ruby-3.3.0/lib/ruby/3.3.0+0/bundler/resolver.rb:331:in `map'
	from /Users/zzak/.rubies/ruby-3.3.0/lib/ruby/3.3.0+0/bundler/resolver.rb:331:in `prepare_dependencies'
	from /Users/zzak/.rubies/ruby-3.3.0/lib/ruby/3.3.0+0/bundler/resolver.rb:51:in `setup_solver'
	from /Users/zzak/.rubies/ruby-3.3.0/lib/ruby/3.3.0+0/bundler/resolver.rb:28:in `start'
	from /Users/zzak/.rubies/ruby-3.3.0/lib/ruby/3.3.0+0/bundler/definition.rb:554:in `start_resolution'
	from /Users/zzak/.rubies/ruby-3.3.0/lib/ruby/3.3.0+0/bundler/definition.rb:289:in `resolve'
	from /Users/zzak/.rubies/ruby-3.3.0/lib/ruby/3.3.0+0/bundler/definition.rb:507:in `materialize'
	from /Users/zzak/.rubies/ruby-3.3.0/lib/ruby/3.3.0+0/bundler/definition.rb:197:in `specs'
	from /Users/zzak/.rubies/ruby-3.3.0/lib/ruby/3.3.0+0/bundler/definition.rb:254:in `specs_for'
	from /Users/zzak/.rubies/ruby-3.3.0/lib/ruby/3.3.0+0/bundler/runtime.rb:18:in `setup'
	from /Users/zzak/.rubies/ruby-3.3.0/lib/ruby/3.3.0+0/bundler.rb:170:in `setup'
	from /Users/zzak/.rubies/ruby-3.3.0/lib/ruby/3.3.0+0/bundler/setup.rb:20:in `block in <top (required)>'
	from /Users/zzak/.rubies/ruby-3.3.0/lib/ruby/3.3.0+0/bundler/ui/shell.rb:159:in `with_level'
	from /Users/zzak/.rubies/ruby-3.3.0/lib/ruby/3.3.0+0/bundler/ui/shell.rb:111:in `silence'
	from /Users/zzak/.rubies/ruby-3.3.0/lib/ruby/3.3.0+0/bundler/setup.rb:20:in `<top (required)>'
	from <internal:/Users/zzak/.rubies/ruby-3.3.0/lib/ruby/3.3.0+0/rubygems/core_ext/kernel_require.rb>:88:in `require'
	from <internal:/Users/zzak/.rubies/ruby-3.3.0/lib/ruby/3.3.0+0/rubygems/core_ext/kernel_require.rb>:88:in `require'

# ...

/Users/zzak/.rubies/ruby-3.3.0/lib/ruby/3.3.0+0/bundler/source/git/git_proxy.rb:315:in `allowed_with_path': The git source https://github.com/rails/rails.git is not yet checked out. Please run `bundle install` before trying to start your application (Bundler::GitError)
	from /Users/zzak/.rubies/ruby-3.3.0/lib/ruby/3.3.0+0/bundler/source/git/git_proxy.rb:65:in `revision'
	from /Users/zzak/.rubies/ruby-3.3.0/lib/ruby/3.3.0+0/bundler/source/git.rb:228:in `revision'
	from /Users/zzak/.rubies/ruby-3.3.0/lib/ruby/3.3.0+0/bundler/source/git.rb:103:in `install_path'
	from /Users/zzak/.rubies/ruby-3.3.0/lib/ruby/3.3.0+0/bundler/source/path.rb:133:in `expanded_path'
	from /Users/zzak/.rubies/ruby-3.3.0/lib/ruby/3.3.0+0/bundler/source/path.rb:174:in `load_spec_files'
	from /Users/zzak/.rubies/ruby-3.3.0/lib/ruby/3.3.0+0/bundler/source/git.rb:205:in `load_spec_files'
	from /Users/zzak/.rubies/ruby-3.3.0/lib/ruby/3.3.0+0/bundler/source/path.rb:107:in `local_specs'
	from /Users/zzak/.rubies/ruby-3.3.0/lib/ruby/3.3.0+0/bundler/source/git.rb:172:in `specs'
	from /Users/zzak/.rubies/ruby-3.3.0/lib/ruby/3.3.0+0/bundler/source.rb:58:in `spec_names'
	from /Users/zzak/.rubies/ruby-3.3.0/lib/ruby/3.3.0+0/bundler/source_map.rb:21:in `block in all_requirements'
	from /Users/zzak/.rubies/ruby-3.3.0/lib/ruby/3.3.0+0/bundler/source_map.rb:20:in `map'
	from /Users/zzak/.rubies/ruby-3.3.0/lib/ruby/3.3.0+0/bundler/source_map.rb:20:in `all_requirements'
	from /Users/zzak/.rubies/ruby-3.3.0/lib/ruby/3.3.0+0/bundler/definition.rb:845:in `source_requirements'
	from /Users/zzak/.rubies/ruby-3.3.0/lib/ruby/3.3.0+0/bundler/definition.rb:497:in `resolution_packages'
	from /Users/zzak/.rubies/ruby-3.3.0/lib/ruby/3.3.0+0/bundler/definition.rb:486:in `resolver'
	from /Users/zzak/.rubies/ruby-3.3.0/lib/ruby/3.3.0+0/bundler/definition.rb:554:in `start_resolution'
	from /Users/zzak/.rubies/ruby-3.3.0/lib/ruby/3.3.0+0/bundler/definition.rb:289:in `resolve'
	from /Users/zzak/.rubies/ruby-3.3.0/lib/ruby/3.3.0+0/bundler/definition.rb:507:in `materialize'
	from /Users/zzak/.rubies/ruby-3.3.0/lib/ruby/3.3.0+0/bundler/definition.rb:197:in `specs'
	from /Users/zzak/.rubies/ruby-3.3.0/lib/ruby/3.3.0+0/bundler/definition.rb:254:in `specs_for'
	from /Users/zzak/.rubies/ruby-3.3.0/lib/ruby/3.3.0+0/bundler/runtime.rb:18:in `setup'
	from /Users/zzak/.rubies/ruby-3.3.0/lib/ruby/3.3.0+0/bundler.rb:170:in `setup'
	from /Users/zzak/.rubies/ruby-3.3.0/lib/ruby/3.3.0+0/bundler/setup.rb:20:in `block in <top (required)>'
	from /Users/zzak/.rubies/ruby-3.3.0/lib/ruby/3.3.0+0/bundler/ui/shell.rb:159:in `with_level'
	from /Users/zzak/.rubies/ruby-3.3.0/lib/ruby/3.3.0+0/bundler/ui/shell.rb:111:in `silence'
	from /Users/zzak/.rubies/ruby-3.3.0/lib/ruby/3.3.0+0/bundler/setup.rb:20:in `<top (required)>'
	from <internal:/Users/zzak/.rubies/ruby-3.3.0/lib/ruby/3.3.0+0/rubygems/core_ext/kernel_require.rb>:88:in `require'
	from <internal:/Users/zzak/.rubies/ruby-3.3.0/lib/ruby/3.3.0+0/rubygems/core_ext/kernel_require.rb>:88:in `require'
```

</details>

<details>
<summary>I've also written a failing test, because if you check `railties/test/fixtures/tmp` while running this test you will see the Gemfile described above, and no lockfile, supposedly because "bundle install" failed silently</summary>

```diff
diff --git a/railties/test/generators/app_generator_test.rb b/railties/test/generators/app_generator_test.rb
index a116c0f867..f6e947fbf8 100644
--- a/railties/test/generators/app_generator_test.rb
+++ b/railties/test/generators/app_generator_test.rb
@@ -691,6 +691,7 @@ def test_generation_runs_bundle_install
     run_generator_instance

     assert_equal 1, @bundle_commands.count("install")
+    assert_file "Gemfile.lock"
   end

   def test_generation_use_original_bundle_environment
```

</details>

### System configuration
**Rails version**: `main`

**Ruby version**: `ruby 3.3.0dev (2023-02-20T01:50:49Z master b326a5f3dd) [arm64-darwin22]`

**Comment by zzak** [comment]
Ok, so there's two issues here.

Actually I made a typo, the correct command is `rails new <path> --dev`.
Though the behavior for where the flag goes is rather surprising.

The fact that the railtie tests are only noisy on my machine, and not CI is questionable. Still that failing test is interesting, I will try it on CI.

**Comment by ghiculescu** [comment]
> Though the behavior for where the flag goes is rather surprising.

I've been stung by this by before too. I think `rails new --dev` (or any other flag, without a path) should error.

**Comment by zzak** [comment]
Fixed by #47444, thanks @ghiculescu!!
